File: MakEliz/backend/routes/accountable.py
from flask import Blueprint, request, jsonify, session
import sys
import os

sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from database import get_connection

accountable_bp = Blueprint("accountable", __name__)

def require_accountable():
    if "employee_id" not in session:
        return False
    return True

# GET: Assets assigned to this Accountable Party
@accountable_bp.route("/my-assets", methods=["GET"])
def get_my_assets():
    if not require_accountable():
        return jsonify({"success": False, "message": "Access denied."}), 403

    employee_id = session.get("employee_id")

    try:
        conn = get_connection()
        cursor = conn.cursor()

        cursor.execute(
            """
            SELECT
                a.ID,
                a.AssetSN,
                a.AssetName,
                d.Name AS Department,
                l.Name AS Location,
                ag.Name AS AssetGroup,
                a.Description,
                a.WarrantyDate
            FROM Assets a
            LEFT JOIN DepartmentLocations dl ON a.DepartmentLocationID = dl.ID
            LEFT JOIN Departments d ON dl.DepartmentID = d.ID
            LEFT JOIN Locations l ON dl.LocationID = l.ID
            LEFT JOIN AssetGroups ag ON a.AssetGroupID = ag.ID
            WHERE a.EmployeeID = ?
            ORDER BY a.AssetName
            """,
            (employee_id,)
        )
        rows = cursor.fetchall()
        conn.close()

        assets = []
        for row in rows:
            assets.append({
                "id": row[0],
                "asset_sn": row[1],
                "asset_name": row[2],
                "department": row[3] or "",
                "location": row[4] or "",
                "asset_group": row[5] or "",
                "description": row[6] or "",
                "warranty_date": str(row[7]) if row[7] else ""
            })

        return jsonify({"success": True, "assets": assets})

    except Exception as e:
        return jsonify({"success": False, "message": "Database error: " + str(e)}), 500

# GET All assignments for this Accountable Party
@accountable_bp.route("/my-assignments", methods=["GET"])
def get_my_assignments():
    if not require_accountable():
        return jsonify({"success": False, "message": "Access denied."}), 403

    employee_id = session.get("employee_id")

    try:
        conn = get_connection()
        cursor = conn.cursor()

        cursor.execute(
            """
            SELECT
                aa.ID,
                a.AssetSN,
                a.AssetName,
                mgr.FirstName + ' ' + mgr.LastName AS AssignedBy,
                aa.AssignedDate,
                aa.Notes,
                aa.Status,
                CASE WHEN wr.ID IS NOT NULL THEN 1 ELSE 0 END AS ReceivedFormFilled,
                CASE WHEN wc.ID IS NOT NULL THEN 1 ELSE 0 END AS CompletionFormFilled
            FROM AssetAssignments aa
            JOIN Assets a ON aa.AssetID = a.ID
            JOIN Employees mgr ON aa.AssignedByEmployeeID = mgr.ID
            LEFT JOIN WorkReceivedForms wr ON wr.AssignmentID = aa.ID
            LEFT JOIN WorkCompletionForms wc ON wc.AssignmentID = aa.ID
            WHERE aa.AssignedToEmployeeID = ?
            ORDER BY aa.AssignedDate DESC
            """,
            (employee_id,)
        )
        rows = cursor.fetchall()
        conn.close()

        assignments = []
        for row in rows:
            assignments.append({
                "id": row[0],
                "asset_sn": row[1],
                "asset_name": row[2],
                "assigned_by": row[3],
                "assigned_date": str(row[4]) if row[4] else "",
                "notes": row[5] or "",
                "status": row[6],
                "received_form_filled": bool(row[7]),
                "completion_form_filled": bool(row[8])
            })

        return jsonify({"success": True, "assignments": assignments})

    except Exception as e:
        return jsonify({"success": False, "message": "Database error: " + str(e)}), 500

# POST: Submit work received form
@accountable_bp.route("/submit-received", methods=["POST"])
def submit_received():
    if not require_accountable():
        return jsonify({"success": False, "message": "Access denied."}), 403

    employee_id = session.get("employee_id")
    data = request.get_json()

    assignment_id = data.get("assignment_id")
    received_date = data.get("received_date")
    condition = data.get("condition_on_receiving", "")
    comments = data.get("comments_on_receiving", "")

    if not assignment_id or not received_date:
        return jsonify({"success": False, "message": "Assignment and received date are required."}), 400

    try:
        conn = get_connection()
        cursor = conn.cursor()

        cursor.execute(
            "SELECT ID FROM WorkReceivedForms WHERE AssignmentID = ?",
            (assignment_id,)
        )
        if cursor.fetchone():
            conn.close()
            return jsonify({"success": False, "message": "You have already submitted a received form for this assignment."}), 409

        cursor.execute(
            """
            INSERT INTO WorkReceivedForms (AssignmentID, ReceivedDate, ReceivedByEmployeeID, ConditionOnReceiving, CommentsOnReceiving)
            VALUES (?, ?, ?, ?, ?)
            """,
            (assignment_id, received_date, employee_id, condition, comments)
        )

        cursor.execute(
            "UPDATE AssetAssignments SET Status = 'In Progress' WHERE ID = ?",
            (assignment_id,)
        )

        conn.commit()
        conn.close()

        return jsonify({"success": True, "message": "Work received form submitted successfully."})

    except Exception as e:
        return jsonify({"success": False, "message": "Database error: " + str(e)}), 500



# POST: Submit work completion form
@accountable_bp.route("/submit-completion", methods=["POST"])
def submit_completion():
    if not require_accountable():
        return jsonify({"success": False, "message": "Access denied."}), 403

    employee_id = session.get("employee_id")
    data = request.get_json()

    assignment_id = data.get("assignment_id")
    completion_date = data.get("completion_date")
    work_done = data.get("work_done_description", "")
    problems = data.get("problems_encountered", "")
    recommended = data.get("recommended_actions", "")
    final_condition = data.get("final_condition", "")

    if not assignment_id or not completion_date or not work_done:
        return jsonify({"success": False, "message": "Assignment, completion date, and work description are required."}), 400

    try:
        conn = get_connection()
        cursor = conn.cursor()

        cursor.execute(
            "SELECT ID FROM WorkReceivedForms WHERE AssignmentID = ?",
            (assignment_id,)
        )
        if not cursor.fetchone():
            conn.close()
            return jsonify({"success": False, "message": "You must submit the work received form before submitting the completion form."}), 400

        cursor.execute(
            "SELECT ID FROM WorkCompletionForms WHERE AssignmentID = ?",
            (assignment_id,)
        )
        if cursor.fetchone():
            conn.close()
            return jsonify({"success": False, "message": "You have already submitted a completion form for this assignment."}), 409

        cursor.execute(
            """
            INSERT INTO WorkCompletionForms
                (AssignmentID, CompletionDate, SubmittedByEmployeeID, WorkDoneDescription, ProblemsEncountered, RecommendedActions, FinalCondition)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            """,
            (assignment_id, completion_date, employee_id, work_done, problems, recommended, final_condition)
        )

        cursor.execute(
            "UPDATE AssetAssignments SET Status = 'Completed' WHERE ID = ?",
            (assignment_id,)
        )

        conn.commit()
        conn.close()

        return jsonify({"success": True, "message": "Completion form submitted to the Manager successfully."})

    except Exception as e:
        return jsonify({"success": False, "message": "Database error: " + str(e)}), 500

File: MakEliz/frontend/js/accountable.js
// ============================================================
// accountable.js
// Handles the Accountable Party Dashboard
// ============================================================

var API_BASE = "http://127.0.0.1:5000/api";

// ============================================================
// On page load
// ============================================================

window.onload = function () {
    checkSession();
};

function checkSession() {
    fetch(API_BASE + "/auth/session", {
        method: "GET",
        credentials: "include"
    })
    .then(function (response) { return response.json(); })
    .then(function (data) {
        if (!data.logged_in || data.role !== "Accountable Party") {
            window.location.href = "login.html";
        } else {
            document.getElementById("accountable-name").textContent = data.name;
            loadMyAssets();
        }
    })
    .catch(function () {
        window.location.href = "login.html";
    });
}

// ============================================================
// Navigation
// ============================================================

function showSection(sectionId, clickedItem) {
    document.querySelectorAll(".page-section").forEach(function (s) {
        s.classList.remove("active");
    });
    document.querySelectorAll(".sidebar-nav li").forEach(function (li) {
        li.classList.remove("active");
    });

    document.getElementById(sectionId).classList.add("active");
    clickedItem.classList.add("active");

    if (sectionId === "section-assets")          { loadMyAssets(); }
    if (sectionId === "section-assignments")     { loadMyAssignments(); }
    if (sectionId === "section-received-form")   { loadAssignmentDropdowns(); }
    if (sectionId === "section-completion-form") { loadAssignmentDropdowns(); }
}

// ============================================================
// Show message helper
// ============================================================

function showMessage(elementId, text, type) {
    var el = document.getElementById(elementId);
    el.className = "message " + type;
    el.textContent = text;
}

// ============================================================
// My Assets (read-only)
// ============================================================

function loadMyAssets() {
    var tbody = document.getElementById("assets-table-body");
    tbody.innerHTML = "<tr><td colspan='7' class='no-data'>Loading...</td></tr>";

    fetch(API_BASE + "/accountable/my-assets", {
        method: "GET",
        credentials: "include"
    })
    .then(function (r) { return r.json(); })
    .then(function (data) {
        if (!data.success || data.assets.length === 0) {
            tbody.innerHTML = "<tr><td colspan='7' class='no-data'>No assets found.</td></tr>";
            return;
        }

        var html = "";
        data.assets.forEach(function (asset) {
            html += "<tr>";
            html += "<td>" + asset.asset_sn + "</td>";
            html += "<td>" + asset.asset_name + "</td>";
            html += "<td>" + (asset.department || "-") + "</td>";
            html += "<td>" + (asset.location || "-") + "</td>";
            html += "<td>" + (asset.asset_group || "-") + "</td>";
            html += "<td>" + (asset.description || "-") + "</td>";
            html += "<td>" + (asset.warranty_date || "-") + "</td>";
            html += "</tr>";
        });
        tbody.innerHTML = html;
    })
    .catch(function () {
        tbody.innerHTML = "<tr><td colspan='7' class='no-data'>Server error.</td></tr>";
    });
}

function loadMyAssignments() {
    var tbody = document.getElementById("my-assignments-table-body");
    tbody.innerHTML = "<tr><td colspan='8' class='no-data'>Loading...</td></tr>";

    fetch(API_BASE + "/accountable/my-assignments", {
        method: "GET",
        credentials: "include"
    })
    .then(function (r) { return r.json(); })
    .then(function (data) {
        if (!data.success || data.assignments.length === 0) {
            tbody.innerHTML = "<tr><td colspan='8' class='no-data'>No assignments found.</td></tr>";
            return;
        }

        var html = "";
        data.assignments.forEach(function (a) {
            var statusClass = a.status === "Completed"   ? "badge-normal"
                            : a.status === "In Progress" ? "badge-high"
                            : "badge-admin";

            var receivedBadge = a.received_form_filled
                ? "<span class='badge badge-normal'>Submitted</span>"
                : "<span class='badge badge-veryhigh'>Pending</span>";

            var completionBadge = a.completion_form_filled
                ? "<span class='badge badge-normal'>Submitted</span>"
                : "<span class='badge badge-veryhigh'>Pending</span>";

            html += "<tr>";
            html += "<td>" + a.asset_sn + "</td>";
            html += "<td>" + a.asset_name + "</td>";
            html += "<td>" + a.assigned_by + "</td>";
            html += "<td>" + a.assigned_date + "</td>";
            html += "<td>" + (a.notes || "-") + "</td>";
            html += "<td><span class='badge " + statusClass + "'>" + a.status + "</span></td>";
            html += "<td>" + receivedBadge + "</td>";
            html += "<td>" + completionBadge + "</td>";
            html += "</tr>";
        });
        tbody.innerHTML = html;
    })
    .catch(function () {
        tbody.innerHTML = "<tr><td colspan='8' class='no-data'>Server error.</td></tr>";
    });
}

function loadAssignmentDropdowns() {
    fetch(API_BASE + "/accountable/my-assignments", {
        method: "GET",
        credentials: "include"
    })
    .then(function (r) { return r.json(); })
    .then(function (data) {
        var receivedSelect    = document.getElementById("received-assignment");
        var completionSelect  = document.getElementById("completion-assignment");

        receivedSelect.innerHTML   = "<option value=''>-- Select Assignment --</option>";
        completionSelect.innerHTML = "<option value=''>-- Select Assignment --</option>";

        if (data.success) {
            data.assignments.forEach(function (a) {
                var label = a.asset_sn + " - " + a.asset_name + " (Assigned: " + a.assigned_date + ")";

                if (!a.received_form_filled) {
                    receivedSelect.innerHTML += "<option value='" + a.id + "'>" + label + "</option>";
                }

                if (a.received_form_filled && !a.completion_form_filled) {
                    completionSelect.innerHTML += "<option value='" + a.id + "'>" + label + "</option>";
                }
            });

            if (receivedSelect.options.length === 1) {
                receivedSelect.innerHTML = "<option value=''>No pending received forms</option>";
            }
            if (completionSelect.options.length === 1) {
                completionSelect.innerHTML = "<option value=''>No pending completion forms</option>";
            }
        }
    })
    .catch(function () {
        console.log("Could not load assignments for dropdowns.");
    });
}

function submitReceivedForm() {
    var assignmentId = document.getElementById("received-assignment").value;
    var date         = document.getElementById("received-date").value;
    var condition    = document.getElementById("received-condition").value.trim();
    var comments     = document.getElementById("received-comments").value.trim();

    if (!assignmentId || !date) {
        showMessage("received-form-message", "Please select an assignment and enter the date received.", "error");
        return;
    }

    fetch(API_BASE + "/accountable/submit-received", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        credentials: "include",
        body: JSON.stringify({
            assignment_id: assignmentId,
            received_date: date,
            condition_on_receiving: condition,
            comments_on_receiving: comments
        })
    })
    .then(function (r) { return r.json(); })
    .then(function (data) {
        if (data.success) {
            showMessage("received-form-message", "Work received form submitted successfully.", "success");
            document.getElementById("received-assignment").value = "";
            document.getElementById("received-date").value = "";
            document.getElementById("received-condition").value = "";
            document.getElementById("received-comments").value = "";
            loadAssignmentDropdowns();
        } else {
            showMessage("received-form-message", data.message, "error");
        }
    })
    .catch(function () {
        showMessage("received-form-message", "Server error. Could not submit form.", "error");
    });
}

function submitCompletionForm() {
    var assignmentId    = document.getElementById("completion-assignment").value;
    var date            = document.getElementById("completion-date").value;
    var workDone        = document.getElementById("completion-work-done").value.trim();
    var problems        = document.getElementById("completion-problems").value.trim();
    var recommended     = document.getElementById("completion-recommended").value.trim();
    var finalCondition  = document.getElementById("completion-final-condition").value.trim();

    if (!assignmentId || !date || !workDone) {
        showMessage("completion-form-message", "Please select an assignment, enter the date, and describe the work done.", "error");
        return;
    }

    fetch(API_BASE + "/accountable/submit-completion", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        credentials: "include",
        body: JSON.stringify({
            assignment_id: assignmentId,
            completion_date: date,
            work_done_description: workDone,
            problems_encountered: problems,
            recommended_actions: recommended,
            final_condition: finalCondition
        })
    })
    .then(function (r) { return r.json(); })
    .then(function (data) {
        if (data.success) {
            showMessage("completion-form-message", "Completion form submitted to the Manager successfully.", "success");
            document.getElementById("completion-assignment").value = "";
            document.getElementById("completion-date").value = "";
            document.getElementById("completion-work-done").value = "";
            document.getElementById("completion-problems").value = "";
            document.getElementById("completion-recommended").value = "";
            document.getElementById("completion-final-condition").value = "";
            loadAssignmentDropdowns();
        } else {
            showMessage("completion-form-message", data.message, "error");
        }
    })
    .catch(function () {
        showMessage("completion-form-message", "Server error. Could not submit form.", "error");
    });
}

function handleLogout() {
    fetch(API_BASE + "/auth/logout", { method: "POST", credentials: "include" })
    .then(function () { window.location.href = "login.html"; })
    .catch(function () { window.location.href = "login.html"; });
}